src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/async_manifest_view.h"
+
+#include "cloud_storage/cache_service.h"
+#include "cloud_storage/logger.h"
+#include "cloud_storage/remote.h"
+#include "cloud_storage/spillover_manifest.h"
+#include "cloud_storage/types.h"
+#include "cloud_storage_clients/types.h"
+#include "config/configuration.h"
+#include "resource_mgmt/io_priority.h"
+#include "utils/human.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/smp.hh>
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/outcome/success_failure.hpp>
+
+#include <exception>
+
+namespace cloud_storage {
+
+static constexpr size_t max_cache_capacity_bytes = 1_GiB;
+
+materialized_manifest_cache::materialized_manifest_cache(
+  size_t capacity_bytes, retry_chain_logger& parent_logger)
+  : _capacity_bytes(capacity_bytes)
+  , _ctxlog(parent_logger)
+  , _sem(max_cache_capacity_bytes) {
+    vassert(
+      capacity_bytes > 0 && capacity_bytes < max_cache_capacity_bytes,
+      "Invalid cache capacity {}, should be non-zero and below 1GiB",
+      capacity_bytes);
+}
+
+ss::future<ss::semaphore_units<>> materialized_manifest_cache::prepare(
```

Comment:
Can concurrent calls to `prepare` lead to a deadlock? Assume two `prepare` calls for oversized manifests.
The first `prepare` will evict everything and then wait for units. If the second `prepare` lands before the first manifest was added to the cache, then it will have nothing to `evict`, so it will end up waiting for units, but without having requested any evictions.

Reply:
> This method only acquires units, so the one that will get the units first (they will acquire in fifo order) will win.

I agree that the first one will win. The issue is that the second one will wait for units without having requested any evictions. If access into the cache is serial that's fine though.